lib/tools/apk-utils.js

```diff
@@ -456,34 +442,87 @@ apkUtilsMethods.installOrUpgrade = async function (appPath, pkg = null, options
-    log.debug(`The installed '${pkg}' package does not require upgrade ('${pkgVersionName}' >= '${apkVersionName}')`);
-    return;
+    log.debug(`The version name of the installed '${pkg}' is greater or equal to the application version name ('${pkgVersionName}' >= '${apkVersionName}')`);
+    return this.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED;
+  }
+
+  log.debug(`The installed '${pkg}' package is older than '${appPath}' (${pkgVersionCode} < ${apkVersionCode} or '${pkgVersionName}' < '${apkVersionName}')'`);
+  return this.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED;
+};
+
+/**
+ * @typedef {Object} InstallOrUpgradeOptions
+ * @property {number} timeout [60000] - The count of milliseconds to wait until the
+ *                                      app is installed.
+ * @property {boolean} allowTestPackages [false] - Set to true in order to allow test
+ *                                                 packages installation.
+ * @property {boolean} useSdcard [false] - Set to true to install the app on sdcard
+ *                                         instead of the device memory.
+ * @property {boolean} grantPermissions [false] - Set to true in order to grant all the
+ *                                                permissions requested in the application's manifest
+ *                                                automatically after the installation is completed
+ *                                                under Android 6+.
+ */
+
+/**
+ * Install the package from the local file system of upgrade it if an older
+ * version of the same package is already installed.
+ *
+ * @param {string} appPath - The full path to the local package.
+ * @param {?string} pkg - The name of the installed package. The method will
+ *                        perform faster if it is set.
+ * @param {?InstallOrUpgradeOptions} options - Set of install options.
+ * @throws {error} If an unexpected error happens during install.
+ */
+apkUtilsMethods.installOrUpgrade = async function (appPath, pkg = null, options = {}) {
+  if (!util.hasValue(options.timeout)) {
+    options.timeout = APK_INSTALL_TIMEOUT;
+  }
+
+  if (!pkg) {
+    const apkInfo = await this.getApkInfo(appPath);
+    pkg = apkInfo.name;
+  }
+
+  const appState = await this.getApplicationInstallState(appPath, pkg);
+  switch (appState) {
+    case this.APP_INSTALL_STATE.NOT_INSTALLED:
+      log.debug(`Installing '${appPath}'`);
+      await this.install(appPath, Object.assign({}, options, {replace: false}));
+      return;
+    case this.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED:
+    case this.APP_INSTALL_STATE.SAME_VERSION_INSTALLED:
+      log.debug(`There is no need to install/upgrade '${appPath}'`);
+      return;
+    case this.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED:
+      log.debug(`Executing upgrade of '${appPath}'`);
+      break;
+    default:
+      log.debug(`The current install state of '${appPath}' is unknown. Installing anyway`);
```

Comment:
Should we call `await this.install(appPath, Object.assign({}, options, {replace: false}));` like NOT_INSTALLED  since the debug log says `Installing anyway`?

Reply:
same comment as above